cmd/ore/aws/upload.go

```diff
@@ -122,6 +122,26 @@ func defaultBucketURL(urlPrefix, imageName, board, file, region string) (*url.UR
 	return s3URL, nil
 }
 
+func createTags(tags []string, res []string) {
+	if len(tags) > 0 {
+		tagMap := make(map[string]string)
```

Comment:
In switching to a split function we've reverted to creating the same `tagMap` multiple times from input.

You can take a couple of different paths to not have to create the map multiple times -- create the tag map and pass it in to `createTags` or move the actual length check + tagging functionality into a lambda inside of `runUpload` (rather than making this `createTags` function) that accepts a list of ids to tag would both work off the top of my head.

@ajeddeloh do you have any preferences on this?

Reply:
Yeah, we should ideally break that out so we're not duplicating work. Since you can loop over 0 length arrays and call length on maps, it should be fine to just move this out.

It may even be ok to not both checking if there are tags to create and blindly call API.CreateTags, but that just wraps the aws go sdk and I don't know if that will be ok with it. If it is, I'd say drop all the conditionals entirely. Hell, if it isn't we could also modify API.CreateTags to just return nil on empty tag lists.

Reply:
I was trying to think of way to not create the `tagMap` multiple times without more code/duplication than in the original PR.

If I move the creation of `tagMap` back into `runUpload`, wouldn't I still need to check `len(uploadTags) > 0` or `len(tagMap) > 0` before calling my `createTags` or `API.CreateTags` for tagging the PV?  Let me know if I'm missing something.

Reply:
You _might_ need to check that. Depends on if the aws go sdk allows you to call their CreateTags with and empty tag list. If it does, then you can just blindly call it. I'm personally not a fan of functions that die when you hand them empty lists/maps/etc, so if that is the case, I think we ought to fix [our wrapper](https://github.com/coreos/mantle/blob/master/platform/api/aws/ec2.go#L199) to just return early if there's nothing to do.

Reply:
You could leave the original instantiation of `tagMap` from the previous PR and then do the following (inside of `runUpload`):

```golang
tagger := func(resources []string) {
	if len(uploadTags) > 0 {
		err = API.CreateTags(resources, tagMap)

		if err != nil {
			fmt.Fprintf(os.Stderr, "unable to add tags: %v\n", err)
			os.Exit(1)
		}
	}
}

...

tagger([]string{hvmID, sourceSnapshot})
```

Reply:
I could wrap the API.CreateTags like this

```
func createTags(t map[string]string, resource []string) {
  if t != nil {
    API.CreateTags(resource, t)
  }
}

// back in runUpload...
var tagMap map[string]string
if len(uploadTags) > 0 {
  // create tagMap
  createTags(tagMap, []string{hvmID, sourceSnapshot})
}

if uploadCreatePV {
  ...
  createTags(tagMap, []string{pvID})
}
```

Reply:
If you were to go with that approach personally I'd move the part checking if the `uploadTags` length & `tagMap` creation onto their own so you could have something like:
```golang
var tagMap map[string]string
if len(uploadTags) > 0 {
    // create tagMap
}

...

createTags(tagMap, []string{hvmID, sourceSnapshot})

...

if uploadCreatePV {
    ...
    createTags(tagMap, []string{pvID})
}
```

But that's a pretty minor stylistic thing on my part that I'd be fine either way with.

----

Overall though I'd lean more towards @ajeddeloh's approach to update the `CreateTags` function inside of the API to handle receiving a nil map (if `aws/aws-sdk-go` can't handle a nil map).

Reply:
Yeah @ajeddeloh's approach seems the cleanest.  In going with that approach, should I just no-op and return nil in API.CreateTags when len(tags) == 0?

Reply:
Yeah that sounds good. That probably also gets rid of the need for the wrapper function here. 

Also, golang has some weird edge cases with maps/slices and nil, so always check `len(things) == 0` not `thing == nil`.

Reply:
If `aws/aws-sdk-go` doesn't properly handle `nil` tag maps then yes just add a check if the tags map is `nil` and if so immediately return `nil`.

Reply:
The docs make it seem like it expects a non-zero-length list, so even if they allow it right now they may break that in the future. 

go map/slice nil weirdness:
https://play.golang.org/p/LRsYY9mHo3c

Reply:
> The docs make it seem like it expects a non-zero-length list, so even if they allow it right now they may break that in the future.
go map/slice nil weirdness:
https://play.golang.org/p/LRsYY9mHo3c

Funny, the instructor for the golang course I just took spent some time on this.  I guess its the difference between declaration and initialization.  

https://blog.golang.org/go-maps-in-action

In the end it makes sense to always check length == 0 unless you are explicitly checking for a map that has been declared but not initialized.